hail/src/main/scala/is/hail/expr/ir/BlockMatrixIR.scala

```diff
@@ -201,22 +206,10 @@ case class BlockMatrixMap2(left: BlockMatrixIR, right: BlockMatrixIR, applyBinOp
   }
 }
 
-object Broadcast2D extends Enumeration {
-  type Broadcast2D = Value
-  val SCALAR, ROW, COL = Value
-
-  def fromString(str: String): Broadcast2D = {
-    str match {
-      case "scalar" => SCALAR
-      case "row" => ROW
-      case "col" => COL
-    }
-  }
-}
-
 case class BlockMatrixBroadcast(
   child: BlockMatrixIR,
-  broadcastKind: Broadcast2D,
+  inIndexExpr: IndexedSeq[Int],
+  outIndexExpr: IndexedSeq[Int],
   shape: IndexedSeq[Long],
```

Comment:
In general, there should be more asserts checking preconditions and invariants. For example, in the broadcast, assuming my suggestion to get rid of `outIndexExpr`, there is a precondition on `shape` that could be expressed `inIndexExpr.zipWithIndex.forall((out, in) => child.shape(in) == shape(out))`.

There's also a precondition on `inIndexExpr`: `inIndexExpr.length < 2 || (inIndexExpr.length == 2 && inIndexExpr(0) != inIndexExpr(1))`.

Right now, we are tracking the shape statically, so these preconditions can all be checked statically, in the constructors of the IR nodes. When we don't know shape statically, we should still keep these assertions, they would just need to move to the `execute` method.

It's a good practice to try to list all the assumptions on the inputs that are required to guarantee the result is correct, and then either document those in comments or (better) in assertions.